### PR TITLE
feat(sequencers): added MsgUpsertSequencer

### PR DESCRIPTION
**Pull request by keruch**
## Description

----

ADR link https://www.notion.so/dymension/ADR-x-Sequencer-Reward-Address-On-Rollapp-da84af6888c141d0a4c1a8df256a5025
Closes https://github.com/dymensionxyz/dymension/issues/1248

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

PR review checkboxes:

I have...

- [ ]  Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ]  Targeted PR against the correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to the GitHub issue with discussion and accepted design
- [ ]  Targets only one GitHub issue
- [ ]  Wrote unit and integration tests
- [ ]  Wrote relevant migration scripts if necessary
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)
- [ ]  Add an issue in the [e2e-tests repo](https://github.com/dymensionxyz/e2e-tests) if necessary

SDK Checklist
- [ ] Import/Export Genesis
- [ ] Registered Invariants
- [ ] Registered Events
- [ ] Updated openapi.yaml
- [ ] No usage of go `map`
- [ ] No usage of `time.Now()`
- [ ] Used fixed point arithmetic and not float arithmetic
- [ ] Avoid panicking in Begin/End block as much as possible
- [ ] No unexpected math Overflow
- [ ] Used `sendCoin` and not `SendCoins`
- [ ] Out-of-block compute is bounded
- [ ] No serialized ID at the end of store keys
- [ ] UInt to byte conversion should use BigEndian

Full security checklist [here](https://www.faulttolerant.xyz/2024-01-16-cosmos-security-1/)

----

For Reviewer:

- [ ]  Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  Confirmed all author checklist items have been addressed

----

After reviewer approval:

- [ ]  In case the PR targets the main branch, PR should not be squash merge in order to keep meaningful git history.
- [ ]  In case the PR targets a release branch, PR must be rebased.

### CODE REVIEW

**Comment by danwt** [comment]
build fail

**Review by zale144 — CHANGES_REQUESTED** [review]

File: utils/uevent/events.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
+// Package uevent is a copy of https://github.com/dymensionxyz/sdk-utils/tree/main/utils/uevent.
+// TODO: import sdk-utils directly when the RDK is updated to SDK v0.47 and further.
```
Comment (mtsitrin) [inline utils/uevent/events.go:2]: we have sdk-util's branch dedicated for the RDK (based on v0.46)

File: x/sequencers/keeper/msg_server.go [hunk 2]
```diff
@@ -43,6 +46,32 @@ func (m msgServer) CreateSequencer(goCtx context.Context, msg *types.MsgCreateSe
 	return &types.MsgCreateSequencerResponse{}, nil
 }
 
+func (m msgServer) UpsertSequencer(goCtx context.Context, msg *types.ConsensusMsgUpsertSequencer) (*types.ConsensusMsgUpsertSequencerResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// all must-methods are safe to use since they're validated in ValidateBasic
+
+	v := msg.MustValidator()
+	m.SetSequencer(ctx, v)
```
Comment (mtsitrin) [inline x/sequencers/keeper/msg_server.go:55]: validate it's not already exists?

File: proto/sequencers/tx.proto [hunk 3]
```diff
@@ -12,6 +11,8 @@ option go_package = "github.com/dymensionxyz/dymension-rdk/x/sequencers/types";
 service Msg {
     rpc CreateSequencer(MsgCreateSequencer) returns (MsgCreateSequencerResponse);
```
Comment (mtsitrin) [inline proto/sequencers/tx.proto:12]: `CreateSequencer` should be removed if there's no use case for it

File: x/sequencers/types/helpers.go [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package types
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/bech32"
+)
+
+func Bech32ToAddr[T sdk.AccAddress | sdk.ValAddress](addr string) (T, error) {
```
Comment (mtsitrin) [inline x/sequencers/types/helpers.go:10]: move to `utils/addressutils`

**Review by mtsitrin — CHANGES_REQUESTED** [review]

File: utils/addressutils/utils.go [hunk 5]
```diff
@@ -27,3 +30,12 @@ func DeriveAddress(addressType AddressType, moduleName, name string) sdk.AccAddr
 		return sdk.AccAddress{}
 	}
 }
+
+// Bech32ToAddr casts an arbitrary-prefixed bech32 string to either sdk.AccAddress or sdk.ValAddress.
+func Bech32ToAddr[T sdk.AccAddress | sdk.ValAddress](addr string) (T, error) {
```
Comment (zale144) [inline utils/addressutils/utils.go:35]: this could maybe be in the `sdk-utils` repo

Comment (zale144) [reply utils/addressutils/utils.go:35]: just saw your comment and copied code below. I agree, we should first migrate to v0.47

**Review by zale144 — APPROVED** [review]
lgtm

**Review by omritoptix — COMMENTED** [review]
is `CreateSequencer` still needed? seems to me it can cause a situation where an operator basically can cause a collision as he may create a sequencer using connecting an already existing consensusAddr to a different bech32 address (vs the default settlement address).

**Review by mtsitrin — APPROVED** [review]

**Pull request merged**